### PR TITLE
gcp_authn: Make gcp auth filter compatibile with gRPC

### PR DESCRIPTION
**Pull request by tyxia**
gcp_auth API is now also used by gRPC dataplane. The API change is from @markdroth in https://github.com/envoyproxy/envoy/pull/35145. This PR combines both API and Implementation. 

Description from @markdroth : "gRPC is implementing support for this filter. However, most gRPC implementations will delegate to the existing [google auth libraries ](https://googleapis.dev/nodejs/google-auth-library/5.6.1/)for the actual token functionality, and those libraries do not allow configuring many of the parameters present in this configuration. This PR updates the config accordingly"

Additional Description : 
Deprecate config.core.v3.HttpUri so [URI](https://github.com/envoyproxy/envoy/blob/3d4e92e2e9d11f3a2c62076bb44fc93e24f878b5/api/envoy/config/core/v3/http_uri.proto#L31) is fixed string rather than configureable option. 

Not breaking change as URI has been [fixed string](https://github.com/envoyproxy/envoy/blob/3d4e92e2e9d11f3a2c62076bb44fc93e24f878b5/api/envoy/extensions/filters/http/gcp_authn/v3/gcp_authn.proto#L26)

### CODE REVIEW

File: source/extensions/filters/http/gcp_authn/gcp_authn_filter.cc [hunk 1]
```diff
@@ -74,8 +74,7 @@ Http::FilterHeadersStatus GcpAuthnFilter::decodeHeaders(Http::RequestHeaderMap&
     // "http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/identity?audience=[AUDIENCE]"
     // So, we add the audience from the config to the final url by substituting the `[AUDIENCE]`
     // with real audience string from the config.
-    std::string final_url =
-        absl::StrReplaceAll(filter_config_->http_uri().uri(), {{"[AUDIENCE]", audience_str_}});
+    std::string final_url = absl::StrReplaceAll(UrlString, {{"[AUDIENCE]", audience_str_}});
```
Comment (yanavlasov) [inline source/extensions/filters/http/gcp_authn/gcp_authn_filter.cc:77]: This change decommissions the field, not deprecates. For example if someone had this URL configured to something other than `UrlString` value, they will get unexpected change in behavior. 

Is it possible for `http_uri` to be different from `UrlString` ? If so you would need to respect user configured value here.

**Review by yanavlasov — COMMENTED** [review]
/wait

File: source/extensions/filters/http/gcp_authn/gcp_authn_filter.cc (hunk 1 above)
Comment (tyxia) [reply source/extensions/filters/http/gcp_authn/gcp_authn_filter.cc:77]: Currently filter actually doesn't support URI different from `UrlString`. So I would not expect user have different configured value here.

But to be safer, I can use runtime guard. The long-term goal is to make this URL fixed string as `UrlString`

**Review by markdroth — APPROVED** [review]
Thanks for doing this!

**Comment by htuch** [comment]
Can you share a link to the relevant libraries in the commit message? I think it's worth maintaining the bread crumbs so we can see why this has changed. Also, why don't we use this library in Envoy @tyxia ?

**Comment by tyxia** [comment]
> Can you share a link to the relevant libraries in the commit message? I think it's worth maintaining the bread crumbs so we can see why this has changed. Also, why don't we use this library in Envoy @tyxia ?

AFAIU, gRPC want to delegate it to the existing google auth libraries specifically for Java and Go. @markdroth Could you please shed some lights here from gRPC side? Is [this](https://googleapis.dev/nodejs/google-auth-library/5.6.1/ ) the library you are using 

(BTW, this filter is only responsible for fetching the token (including decoding and caching), the token authentication is handled by receiving service)

**Comment by markdroth** [comment]
> AFAIU, gRPC want to delegate it to the existing google auth libraries specifically for Java and Go. @markdroth Could you please shed some lights here from gRPC side?

Yes, that's correct, this is an issue for Java, Go, Python, and Node, since those are the languages that have google auth libraries.  See https://cloud.google.com/run/docs/authenticating/service-to-service#use_the_authentication_libraries for some examples of how these libraries would be used.

> Also, why don't we use this library in Envoy @tyxia ?

Unfortunately, C++ does not have a google auth library implementation, so Envoy needed to implement this functionality itself, and gRPC C++ will need to do the same thing.  However, I don't think we want to be in a place where the xDS API surface exposes options that are fundamentally incompatible with what the google auth libraries support.

**Comment by htuch** [comment]
The main reason I'm asking is I'm curious if these libraries support the ability to override metadata URI when used in places like our staging or other test environments. 100% clear on this not being a normal operation mode concern.

**Comment by yanavlasov** [comment]
My concern is that may have to change it again soon to support JWT token minting in CSM. It may require a different URL or additional parameters to the query.

**Review by yanavlasov — APPROVED** [review]

**Comment by tyxia** [comment]
> My concern is that may have to change it again soon to support JWT token minting in CSM. It may require a different URL or additional parameters to the query.

Exactly,  I mentioned the same thing in the internal email with CSM. I am thinking if the final design after internal discussion requires the URL or some other things, we can add a new field to support that.

I will loop in gRPC team as well

**Comment by markdroth** [comment]
@tyxia, can I suggest that you sync up with the google auth libraries team on what the semantics should be here?  If there is an option to override the URI for (e.g.) staging use, then maybe the semantic we want here is to make it an optional field, defaulting to the normal URI but allowing us to override it.

If the auth libraries do not have an option to override the URI but we wind up needing a different URI for a different use-case, then I assume the google auth libraries would expose that as a different kind of token provider.  I think we would want to do the same thing in xDS, representing it as a different filter.  Note that if we do that, we can still share a lot of code between the two filters; we just need the xDS API surface to map closely to the google auth library API.

**Comment by tyxia** [comment]
@markdroth The CSM discussion is still in early stage and it requires some clarification and discussion. I am ok with this PR and looks like @yanavlasov and @htuch have both approved it. 

Yes, I will sync up with google auth and you (if any API change is needed) once when we have a clearer picture on CSM feature.

**Comment by markdroth** [comment]
Sounds good.  Thank you!

**Pull request merged**